### PR TITLE
[Execution] Add support for configurable transaction filtering

### PR DESCRIPTION
**Pull request by sitalkedia**
### Description

A filter that can be used to allow or deny transactions from being executed. It contains a set
of rules that are evaluated one by one in the order of declaration. If a rule matches, the transaction is either allowed or
denied depending on the rule. If no rule matches, the transaction is allowed. For example a rules might look like this:
```
    rules:
        - Allow:
            Sender: f8871acf2c827d40e23b71f6ff2b9accef8dbb17709b88bd9eb95e6bb748c25a
        - Allow:
            ModuleAddress: "0000000000000000000000000000000000000000000000000000000000000001"
        - Allow:
            EntryFunction:
                - "0000000000000000000000000000000000000000000000000000000000000001"
                - test
                - check
        - Allow:
            EntryFunction:
                - "0000000000000000000000000000000000000000000000000000000000000001"
                - test
                - new
        - Deny: All
 ```
This filter allows transactions from the sender with address f8871acf2c827d40e23b71f6ff2b9accef8dbb17709b88bd9eb95e6bb748c25a or
from the module with address 0000000000000000000000000000000000000000000000000000000000000001 or entry functions
test::check and test::new from the module 0000000000000000000000000000000000000000000000000000000000000001. All other transactions are denied.

Support for various types of rules have been added as below 

```
    All,
    BlockId(HashValue),
    TransactionId(HashValue),
    Sender(AccountAddress),
    ModuleAddress(AccountAddress),
    EntryFunction(AccountAddress, String, String),
 ```

### Test Plan

Added comprehensive UTs.

### CODE REVIEW

**Comment by msmouse** [comment]
why not a ACL -like solution, like
``` yaml
txn_acl:
   - DenyBlockId: xx
   - DenyTxnHash: xx
   - AllowSender: xx
   - DenySender: xx
   - AllowModuleAuthor: 0x1
   - DenyModule: 0xCafe::Graphio
   - DenyEntryFunction: 0xCafe::Graphio::draw
   - DenyAll
```

^ rules are evaluated one by one, with Allow as default, so an empty list results in all txns / blocks being allowed.

**Comment by sitalkedia** [comment]
> why not a ACL -like solution, like
> 
> ```yaml
> txn_acl:
>    - DenyBlockId: xx
>    - DenyTxnHash: xx
>    - AllowSender: xx
>    - DenySender: xx
>    - AllowModuleAuthor: 0x1
>    - DenyModule: 0xCafe::Graphio
>    - DenyEntryFunction: 0xCafe::Graphio::draw
>    - DenyAll
> ```
> 
> ^ rules are evaluated one by one, with Allow as default, so an empty list results in all txns / blocks being allowed.

How would this work if we have both Allow and Deny filter for same (EntryFunction for example) enabled?

**Comment by msmouse** [comment]
@sitalkedia  return on the first rule matched

**Comment by sitalkedia** [comment]
> @sitalkedia return on the first rule matched

That IMO is going to create confusion around how to configure this filter.. in practice I don't see a need to have chaining of filters.. but if you feel strongly, I can add that support.

**Comment by msmouse** [comment]
Yeah.. 
I think it's pretty intuitive to many. For example: http://nginx.org/en/docs/http/ngx_http_access_module.html
And it's straightforward to implement and read: Do a `trait FilterRule` and a `struct TxnFilter(rules: Vec<Arc<dyn FilterRule>>)`

**Comment by msmouse** [comment]
oh.. actually 
``` Rust
enum Matcher {
     ModuleAuthor(AccountAddress),
     TransactionHash(HashValue),
}

enum Rule {
    Allow(Matcher),
    Deny(Matcher),
}

struct Filters {
    rules: Vec<Rule>
}
```

and config:
```
transaction_filters:
    - Allow:
       MoudleAuthor: 0x1235 
```

**Comment by sitalkedia** [comment]
> ```rust
>      ModuleAuthor(AccountAddress),
> ```

Would you prefer a Matcher to look like ` ModuleAuthor(AccountAddress)` or `ModuleAuthor(HashSet<AccountAddress>),. In the first case we can still specify multiple account address by chaining and in the second case no chaining is needed.. The first case will be a bit inefficient for filtering though as we have to iterate though multiple filters..

**Comment by sitalkedia** [comment]
> prototype: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bc3000f14482344edc065506d6c2ac98
> 
> > The first case will be a bit inefficient for filtering though as we have to iterate though multiple filters..
> 
> I'm not sure iterating through the HashSet entries is more efficient? Plus I love the simplicity..

Not iterating though the hashset entries ofcourse, just a simple look up will suffice.. but let's stick to simplicity :)

**Comment by grao1991** [comment]
> > prototype: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bc3000f14482344edc065506d6c2ac98
> > > The first case will be a bit inefficient for filtering though as we have to iterate though multiple filters..
> > 
> > 
> > I'm not sure iterating through the HashSet entries is more efficient? Plus I love the simplicity..
> 
> Not iterating though the hashset entries ofcourse, just a simple look up will suffice.. but let's stick to simplicity :)

I'm in favor of chaining filter rules, there won't be too many items for the same rule, and splitting them into multiple entries will make it easier for monitoring them separately.

**Comment by msmouse** [comment]
supporting `And` should be easy too https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66da037ed5169e1eb7bfdbb5f090d219

**Comment by sitalkedia** [comment]
@msmouse , @grao1991 - Updated as per discussed offline, PTAL.

File: config/src/config/transaction_filter_type.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+enum MatchResult {
+    Allow,
+    Deny,
+    NoMatch,
+}
+
+impl Rule {
+    fn matches(&self, block_id: HashValue, txn: &SignedTransaction) -> MatchResult {
```
Comment (msmouse) [inline config/src/config/transaction_filter_type.rs:60]: nit: name: `fn eval() -> EvalResult`

File: config/src/config/transaction_filter_type.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    ) -> Self {
+        self.rules.push(Rule::Allow(Matcher::EntryFunction(
+            address,
+            module_name,
+            function,
+        )));
+        self
+    }
+
+    pub fn matches(&self, block_id: HashValue, txn: &SignedTransaction) -> bool {
```
Comment (msmouse) [inline config/src/config/transaction_filter_type.rs:182]: nit: name: `fn allows()`

File: consensus/src/transaction_filter/mod.rs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+impl TransactionFilter {
+    pub(crate) fn new(filter: Filter) -> Self {
+        Self { filter }
+    }
+
+    pub fn filter(
+        &self,
+        _block_id: HashValue,
```
Comment (msmouse) [inline consensus/src/transaction_filter/mod.rs:20]: nit: remove underscore.

**Review by msmouse — APPROVED** [review]
Beautiful, boss!

File: config/src/config/transaction_filter_type.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
```
Comment (JoshLind) [inline config/src/config/transaction_filter_type.rs:2]: nit: given this is a new file, we don't need to mention this line about Meta (same below). 😄 But, doesn't matter too much!

File: config/src/config/transaction_filter_type.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+enum Matcher {
+    All,
+    BlockId(HashValue),
+    TransactionId(HashValue),
+    Sender(AccountAddress),
+    ModuleAddress(AccountAddress),
```
Comment (JoshLind) [inline config/src/config/transaction_filter_type.rs:18]: Silly question: Do we want the ability to disable transaction payload types? e.g., disable all transactions of type `TransactionPayload::Script(...)`, `TransactionPayload::ModuleBundle(...)` or `TransactionPayload::Multisig(...)`?

The move experts will probably have to confirm/deny, but I'm just trying to think about edge-cases that may be exploitable through those transaction types and if a filter will be useful.

'cc @davidiw @movekevin, any thoughts? 😄

File: config/src/config/transaction_filter_type.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            module_name,
+            function,
+        )));
+        self
+    }
+
+    pub fn allows(&self, block_id: HashValue, txn: &SignedTransaction) -> bool {
+        for rule in &self.rules {
+            match rule.eval(block_id, txn) {
+                EvalResult::Allow => return true,
```
Comment (JoshLind) [inline config/src/config/transaction_filter_type.rs:185]: nit: Might be worth adding an comment to this method that the first rule match is what is chosen 😄

**Review by JoshLind — COMMENTED** [review]
Looks solid, thanks @sitalkedia 😄 Will wait for the experts to do the final approval.

**Review by grao1991 — APPROVED** [review]

File: config/src/config/transaction_filter_type.rs (hunk 6 above)
Comment (sitalkedia) [reply config/src/config/transaction_filter_type.rs:185]: I have added some comment at the top to make it clear but no harm in adding it again here.

File: config/src/config/transaction_filter_type.rs (hunk 5 above)
Comment (sitalkedia) [reply config/src/config/transaction_filter_type.rs:18]: We can extend this framework further to allow more filters..if needed, may be we can do that as a followup?

**Pull request merged**